lib/AST/Decl.cpp

```diff
@@ -4087,46 +4087,152 @@ bool ProtocolDecl::isAvailableInExistential(const ValueDecl *decl) const {
+        auto *assocDecl = assoc->getAssocType();
+        if (map.count(assocDecl))
+          map[assocDecl] = true;
+      }
+    }
+  }
+}
+
+ExistentialSupportKind
+ProtocolDecl::existentialTypeSupportedSlow(LazyResolver *resolver,
+                                           AssociatedTypeFreedomMap &map,
+                                           bool needsCompleteAssocMap) {
+  if (Bits.ProtocolDecl.ExistentialTypeSupportedValid) {
+    auto supportKd = getExistentialSupportKind();
+    // This function recurses up the protocol hierarchy. If not
+    // supported due to free associated type member, we must let the
+    // inheriting protocol know about it.
+    if (supportKd == ExistentialSupportKind::UnsupportedAssoc ||
+        needsCompleteAssocMap)
+      getAssociatedTypeFreedomMapRec(map);
+    return supportKd;
+  }
+
+  auto setSupportKind = [&](ExistentialSupportKind ESK) {
+    Bits.ProtocolDecl.ExistentialTypeSupported = static_cast<unsigned>(ESK);
+  };
   // Assume for now that the existential type is supported; this
   // prevents circularity issues.
   Bits.ProtocolDecl.ExistentialTypeSupportedValid = true;
-  Bits.ProtocolDecl.ExistentialTypeSupported = true;
+  setSupportKind(ExistentialSupportKind::Supported);
 
-  // ObjC protocols can always be existential.
+  // ObjC protocols can always be existential and never have
+  // associated type members.
   if (isObjC())
-    return true;
+    return ExistentialSupportKind::Supported;
 
+  bool hasAssoc = false;
   for (auto member : getMembers()) {
-    // Check for associated types.
-    if (isa<AssociatedTypeDecl>(member)) {
-      // An existential type cannot be used if the protocol has an
-      // associated type.
-      Bits.ProtocolDecl.ExistentialTypeSupported = false;
-      return false;
+    if (auto assocDecl = dyn_cast<AssociatedTypeDecl>(member)) {
+      // Mark that we found an associated type and collect it.
+      hasAssoc = true;
+      map[assocDecl] = /*isConstrainedToConrete*/ false;
+      continue;
     }
 
-    // For value members, look at their type signatures.
+    // For value members, look for Self references in their type signatures.
     if (auto valueMember = dyn_cast<ValueDecl>(member)) {
       if (resolver && !valueMember->hasInterfaceType())
         resolver->resolveDeclSignature(valueMember);
 
-      if (!isAvailableInExistential(valueMember)) {
-        Bits.ProtocolDecl.ExistentialTypeSupported = false;
-        return false;
+      auto selfKind = findProtocolSelfReferences(
+                        valueMember, /*allowCovariantParameters=*/true,
+                        /*skipAssocTypes=*/true);
+
+      // A protocol with a Self reference in contravariant position
+      // cannot be used as an existential, alongside any inheriting
+      // protocol; compute the rest of the map if needed and bail out.
+      if (selfKind.parameter || selfKind.other) {
+        if (needsCompleteAssocMap)
```

Comment:
It would be better to compute the map as needed and cache it in the protocol, instead of passing this boolean down here.

Reply:
Caching is a good idea, since we only ever need fully computed maps for composition types, but isn't caching orthogonal to that boolean? It's purpose is to simply optimize the computation of the map. I do agree it looks a bit odd though.

Fortunately I found another rather trivial reason to always compute maps and support kinds separately (and thus get rid of that boolean) - if all protocols of a composition type are individually supported, we don't need maps to begin with.